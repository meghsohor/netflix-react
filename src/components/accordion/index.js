import React, { createContext, useContext, useState } from 'react';
import { Container, Inner, Title, Item, Header, Body } from './styles/accordion';

const toggleContext = createContext();

export default function Accordion({ children, ...restProps }) {
  return (
    <Container {...restProps}>
      <Inner>{children}</Inner>
    </Container>
  );
}

Accordion.Title = function AccordionTitle({ children, ...restProps }) {
  return <Title {...restProps}>{children}</Title>;
};


Accordion.Item = function AccordionItem({ children, ...restProps }) {
  const [toggleShow, setToggleShow] = useState(false);
  return (
    <toggleContext.Provider value={{ toggleShow, setToggleShow }}>
      <Item {...restProps}>{children}</Item>
    </toggleContext.Provider>
  );
};

Accordion.Header = function AccordionHeader({ children, ...restProps }) {
  const { toggleShow, setToggleShow } = useContext(toggleContext);

  return (
    <Header onClick={() => setToggleShow((toggleShow) => !toggleShow)} {...restProps}>
      {children}
      {/* We can check a state in following way */}
      {/* <pre>{JSON.stringify(toggleShow)}</pre> */}
      {toggleShow ? (
        <img src="/images/icons/close-slim.png" alt="Close" />
      ) : (
        <img src="/images/icons/add.png" alt="Open" />
      )}
    </Header>
  );
};

Accordion.Body = function AccordionBody({ children, ...restProps }) {
  const { toggleShow } = useContext(toggleContext);

  /* return toggleShow ? <Body {...restProps}>{children}</Body> : null; */
  return (
    <Body className={toggleShow ? 'open' : 'closed'} {...restProps}>
      <span>{children}</span>
    </Body>
  );
};
