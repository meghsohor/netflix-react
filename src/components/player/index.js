import React, { useState, useContext, createContext } from 'react';
import ReactDOM from 'react-dom';
import { Container, Button, Outer, Overlay, Inner, Close } from './styles/player';

export const PlayerContext = createContext();

export default function Player({ children, ...restProps }) {
  const [showPlayer, setShowPlayer] = useState(false);

  return (
    <PlayerContext.Provider value={{ showPlayer, setShowPlayer }}>
      <Container {...restProps}>{children}</Container>
    </PlayerContext.Provider>
  );
}

Player.Video = function PlayerVideo({ src, ...restProps }) {
  const { showPlayer, setShowPlayer } = useContext(PlayerContext);

  return showPlayer
    ? ReactDOM.createPortal(
        <Outer data-testid="player" {...restProps}>
          <Overlay onClick={() => setShowPlayer(false)} data-testid="player-overlay" />
          <Inner>
            <video id="netflix-player" controls>
              <source src={src} type="video/mp4" />
            </video>
            <Close onClick={() => setShowPlayer(false)} data-testid="player-close" />
          </Inner>
        </Outer>,
        document.body
      )
    : null;
};

Player.Button = function PlayerButton({ ...restProps }) {
  const { showPlayer, setShowPlayer } = useContext(PlayerContext);

  return (
    <Button onClick={() => setShowPlayer((showPlayer) => !showPlayer)} {...restProps}>
      Play
    </Button>
  );
};
