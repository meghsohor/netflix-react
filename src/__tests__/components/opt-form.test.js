import React from 'react';
import { render } from '@testing-library/react';
import { OptForm } from '../../components';

describe('<OptForm />', () => {
  it('renders the <OptForm /> with populated data', () => {
    const { container, getByText, getByPlaceholderText } = render(
      <OptForm>
        <OptForm.Text>Ready to watch? Enter your email to create or restart your membership.</OptForm.Text>
        <OptForm.Break />
        <OptForm.Input type="email" placeholder="Email address" />
        <OptForm.Button>Get Started</OptForm.Button>
      </OptForm>
    );

    expect(getByText('Get Started')).toBeTruthy();
    expect(getByText('Ready to watch? Enter your email to create or restart your membership.')).toBeTruthy();
    expect(getByPlaceholderText('Email address')).toBeTruthy();
    expect(container.firstChild).toMatchSnapshot();
  });
});
