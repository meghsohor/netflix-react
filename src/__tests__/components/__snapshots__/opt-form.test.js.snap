// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OptForm /> renders the <OptForm /> with populated data 1`] = `
<div
  class="opt-form__Container-sc-8oudnm-0 bhlYJp"
>
  <p
    class="opt-form__Text-sc-8oudnm-3 loEVEx"
  >
    Ready to watch? Enter your email to create or restart your membership.
  </p>
  <div
    class="opt-form__Break-sc-8oudnm-4 hHQsws"
  />
  <input
    class="opt-form__Input-sc-8oudnm-1 btBYpX"
    placeholder="Email address"
    type="email"
  />
  <button
    class="opt-form__Button-sc-8oudnm-2 eZRfuW"
  >
    Get Started
    <img
      alt="Try Now"
      src="/images/icons/chevron-right.png"
    />
  </button>
</div>
`;
